extensions/guacamole-auth-cas/src/main/java/org/apache/guacamole/auth/cas/ticket/TicketValidationService.java

```diff
@@ -88,7 +92,7 @@
      *     If the ID ticket is not valid or guacamole.properties could
      *     not be parsed.
      */
-    public Map<String, String> validateTicket(String ticket,
+    public TokensAndGroups  validateTicket(String ticket,
```

Comment:
I hate to be a pain about this, but I'm still in favor of having this `validateTicket()` method just return the `Map<String, String>` data, and parsing it out in the body of the `authenticateUser()` method in the `AutenticationProviderService` class.  This leaves the `validateTicket()` method to do ticket validation, keeping it free from the extra logic introduced here.

I'd suggest doing one of two things:
* Keep the `TokenAndGroups` utility class, but, rather than using it to store the information, implement a static method (or two) that takes in the `Map<String, String>` input from `validateTicket()` and parses out the groups, returning the `List` of effective groups.  Call this method within the `authenticateUser()` method using the output received from `validateTicket()` and using the output to set up the effective groups
* Get rid of the `TokensAndGroups` class and implement the parser in a `private` method within the `AuthenticationProviderService` class that does the same as above.

I'm completely on-board with leaving the group data in the tokens, as we discussed earlier.

Reply:
Ok, the problem with your suggestion is that the `Map<String, String>` **_presently_** returned by `validateTicket()` doesn't contain valid information for me to derive groups.  In particular, it returns capitalized keys with a string prepended (CAS-FOOBAR), and groups is case sensitive (So, CAS-FOOBAR, would not be enough for me to differentiate groups such as FooBar from fooBar - I had mentioned this before).  So, to do the processing within `authenticateUser()`, I would need to:

- Refactor the previous author's code to return what I need ("raw" SAML attributes released by CAS)
- Move/restructure their code to create tokens from SAML attributes to `authenticateUser()`
- Derive both tokens and groups in `authenticateUser()`

Certainly possible, but refactoring/restructuring someone elses code (and then having to QA it) does seem like a different task and a bit more risk than adding a feature to existing code.   

Also, does that really fit into the philosophy of "one jira ticket, one change?" or is it "ok" to "fix" someone elses non-broken code as part of adding a feature?  I'm not even a Java programmer and I can see a lot of things I'd like to "fix" given a license like that...  :-)

I'll do it (given the time to do so), if that's what you'd like to see.

Reply:
> ...
>
> Certainly possible, but refactoring/restructuring someone elses code (and then having to QA it) does seem like a different task and a bit more risk than adding a feature to existing code.
>
> Also, does that really fit into the philosophy of "one jira ticket, one change?" or is it "ok" to "fix" someone elses non-broken code as part of adding a feature? ...

It does occasionally happen that additional code needs to be touched in order to implement a feature. I don't think there's a universal "this way is safe, this way is unsafe" answer to your question, but we do tend to look for ways to improve established code when a new change indicates that it would be a better direction. Doing otherwise would mean we trend toward greater and greater technical debt.